File: sunfounder/04_PWM/led_pwm.py
import RPi.GPIO as GPIO
import time

LED_PIN = 12

def setup():
    global p
    GPIO.setmode(GPIO.BOARD)  # physical location
    GPIO.setup(LED_PIN, GPIO.OUT)
    led_off();
    p = GPIO.PWM(LED_PIN, 1000) # frequency = 1khz
    p.start(0)
    
def led_off():
    GPIO.output(LED_PIN, GPIO.LOW)

def led_on():
    GPIO.output(LED_PIN, GPIO.HIGH)
    
def destroy():
    led_off()
    GPIO.cleanup()
    
def loop():
    while True:
        for dc in range(0,101,4):
            p.ChangeDutyCycle(dc)
            time.sleep(0.02)
        #time.sleep(1)
        for dc in range(100,-1,-4):
            p.ChangeDutyCycle(dc)
            time.sleep(0.02)
        #time.sleep(1)
    
if __name__ == '__main__':
    setup()
    loop()
    destroy()
    

File: sunfounder/04_PWM/led_pwm.c
#include <wiringPi.h>
#include <stdlib.h>
#include <stdio.h>

#define LED_PIN 1

int led_pwm(int dutyCycle) {
	pwmWrite(LED_PIN, dutyCycle);
	delay(2);
}

int led_on() {
	digitalWrite(LED_PIN, HIGH);
}

int led_off() {
	digitalWrite(LED_PIN, LOW);
}

int init() {
	printf("setting up...\n");
	if (wiringPiSetup()==-1) {
		printf("setup wiringPi failed\n");
		return -1;
	}
	pinMode(LED_PIN, PWM_OUTPUT);
	led_off();
	return 0;
}

int destroy() {
	printf("cleaning up...\n");
	return 0;
}

int loop() {
	int i;
	while (1) {
		for (i=0; i<1024; i++) {
			led_pwm(i);
		}
		for (i=1023; i>=0; i--) {
			led_pwm(i);
		}
	}
	return 0;
}

int main(int argc, char* argv[]) {
	init();
	loop();
	destroy();
}